crates/autopilot/src/decoded_settlement.rs

```diff
@@ -0,0 +1,487 @@
+    // Tricky because the decoded settlement is using FILLED `orders` so we don't
+    // always know the executed amount in case of partial fill.
+    pub fn total_fees(
+        &self,
+        external_prices: &ExternalPrices,
+        orders: &[Order],
+        configuration: &FeeConfiguration,
+    ) -> U256 {
+        self.trades.iter().fold(0.into(), |acc, trade| {
+            match orders
+                .iter()
+                .find(|order| order.signature == trade.signature.0)
+            {
+                Some(order) => {
+                    acc + match fee(trade, external_prices, order, configuration) {
+                        Some(fee) => fee,
+                        None => {
+                            tracing::warn!("possible incomplete fee calculation");
+                            0.into()
+                        }
+                    }
+                }
+                None => {
+                    tracing::warn!(
+                        "order not found for trade, possible incomplete fee calculation"
+                    );
+                    acc
+                }
+            }
+        })
+    }
+}
+
+fn surplus(
+    trade: &DecodedTrade,
+    tokens: &[Address],
+    clearing_prices: &[U256],
+    external_prices: &ExternalPrices,
+) -> Option<U256> {
+    let sell_token_index = trade.sell_token_index.as_u64() as usize;
+    let buy_token_index = trade.buy_token_index.as_u64() as usize;
+
+    let sell_token_clearing_price = clearing_prices.get(sell_token_index)?.to_big_rational();
+    let buy_token_clearing_price = clearing_prices.get(buy_token_index)?.to_big_rational();
+    let kind = order_kind(&trade.flags).unwrap();
+
+    if match kind {
+        OrderKind::Sell => &buy_token_clearing_price,
+        OrderKind::Buy => &sell_token_clearing_price,
+    }
+    .is_zero()
+    {
+        return None;
+    }
+
+    let surplus = trade_surplus(
+        kind,
+        &trade.sell_amount.to_big_rational(),
+        &trade.buy_amount.to_big_rational(),
+        &trade.executed_amount.to_big_rational(),
+        &sell_token_clearing_price,
+        &buy_token_clearing_price,
+    )?;
+
+    let normalized_surplus = match kind {
+        OrderKind::Sell => {
+            let buy_token = tokens.get(buy_token_index)?;
+            external_prices.get_native_amount(*buy_token, surplus / buy_token_clearing_price)
+        }
+        OrderKind::Buy => {
+            let sell_token = tokens.get(sell_token_index)?;
+            external_prices.get_native_amount(*sell_token, surplus / sell_token_clearing_price)
+        }
+    };
+
+    big_rational_to_u256(&normalized_surplus).ok()
+}
+
+fn fee(
+    trade: &DecodedTrade,
+    external_prices: &ExternalPrices,
+    order: &Order,
+    configuration: &FeeConfiguration,
+) -> Option<U256> {
+    let scaled_fee_amount =
+        order.full_fee_amount * U256::from_f64_lossy(configuration.fee_objective_scaling_factor);
```

Comment:
Turning the `fee_objective_scaling_factor` directly into a `U256` will lose precision as we will probably use very small values (maybe `[1.0, 2.0]`). I think the correct thing would be to turn `full_fee_amount` and `fee_objective_scaling_factor` into `BigRational`, compute the total fee and only then turn the result back into a `U256`.
And in that case you could store `fee_objective_scaling_factor` directly as a `BigRational` in the `FeeConfiguration` to only convert it once.
To have total consistency we could store the `fee_objective_scaling_factor` together with the solutions when we determined the winner.

Reply:
Ah nice catch, did not test this properly. Thanks.

Reply:
I think since this code is important to compute correct rewards we should probably not take any unnecessary shortcuts and avoid loss of precision wherever possible. The code using all `BigRational` is not even more verbose/uglier IMO:
```rust
    let full_fee_amount = u256_to_big_rational(&order.full_fee_amount);
    let scaling_factor = BigRational::from_f64(configuration.fee_objective_scaling_factor)?;
    let scaled_fee_amount = full_fee_amount * scaling_factor;

    let fee = match order.kind {
        model::order::OrderKind::Buy => {
            scaled_fee_amount * u256_to_big_rational(&trade.executed_amount)
                / u256_to_big_rational(&trade.buy_amount)
        }
        model::order::OrderKind::Sell => {
            scaled_fee_amount * u256_to_big_rational(&trade.executed_amount)
                / u256_to_big_rational(&trade.sell_amount)
        }
    };

    let fee = external_prices.try_get_native_amount(order.sell_token, fee)?;
    big_rational_to_u256(&fee).ok()
```

Reply:
Updated. Set fee_objective_scaling_factor to be BigRational